archive/tar_stream_test.ts

```diff
@@ -0,0 +1,302 @@
+import {
+  parsePathname,
+  TarStream,
+  type TarStreamInput,
+  validTarStreamOptions,
+} from "./tar_stream.ts";
+import { assertEquals } from "../assert/mod.ts";
+import { assert } from "../assert/assert.ts";
+
+Deno.test("TarStream() with default stream", async () => {
+  const text = new TextEncoder().encode("Hello World!");
+
+  const reader = ReadableStream.from<TarStreamInput>([
+    {
+      pathname: "./potato",
+    },
+    {
+      pathname: "./text.txt",
+      size: text.length,
+      readable: ReadableStream.from([text.slice()]),
+    },
+  ])
+    .pipeThrough(new TarStream())
+    .getReader();
+
+  let size = 0;
+  while (true) {
+    const { done, value } = await reader.read();
+    if (done) {
+      break;
+    }
+    size += value.length;
+  }
+  assertEquals(size, 512 + 512 + Math.ceil(text.length / 512) * 512 + 1024);
+});
+
+Deno.test("TarStream() with byte stream", async () => {
+  const text = new TextEncoder().encode("Hello World!");
+
+  const reader = ReadableStream.from<TarStreamInput>([
+    {
+      pathname: "./potato",
+    },
+    {
+      pathname: "./text.txt",
+      size: text.length,
+      readable: ReadableStream.from([text.slice()]),
+    },
+  ])
+    .pipeThrough(new TarStream())
+    .getReader({ mode: "byob" });
+
+  let size = 0;
+  while (true) {
+    const { done, value } = await reader.read(
+      new Uint8Array(Math.ceil(Math.random() * 1024)),
+    );
+    if (done) {
+      break;
+    }
+    size += value.length;
+  }
+  assertEquals(size, 512 + 512 + Math.ceil(text.length / 512) * 512 + 1024);
+});
+
+Deno.test("TarStream() with negative size", async () => {
+  const text = new TextEncoder().encode("Hello World");
+
+  const readable = ReadableStream.from<TarStreamInput>([
+    {
+      pathname: "name",
+      size: -text.length,
+      readable: ReadableStream.from([text.slice()]),
+    },
+  ])
+    .pipeThrough(new TarStream());
+
+  let threw = false;
+  try {
+    await Array.fromAsync(readable);
+  } catch (error) {
+    threw = true;
+    assert(typeof error === "string");
+    assertEquals(error, "Size cannot exceed 64 Gibs");
+  }
+  assertEquals(threw, true);
+});
+
+Deno.test("TarStream() with 65 GiB size", async () => {
+  const size = 1024 ** 3 * 65;
+  const step = 1024; // Size must equally be divisible by step
+  const iterable = function* () {
+    for (let i = 0; i < size; i += step) {
+      yield new Uint8Array(step).map(() => Math.floor(Math.random() * 256));
+    }
+  }();
+
+  const readable = ReadableStream.from<TarStreamInput>([
+    {
+      pathname: "name",
+      size,
+      readable: ReadableStream.from(iterable),
+    },
+  ])
+    .pipeThrough(new TarStream());
+
+  let threw = false;
+  try {
+    await Array.fromAsync(readable);
+  } catch (error) {
+    threw = true;
+    assert(typeof error === "string");
+    assertEquals(error, "Size cannot exceed 64 Gibs");
+  }
+  assertEquals(threw, true);
```

Comment:
assertRejects only validates that an error was thrown but not that it was the error one is expecting to be thrown

Reply:
Sorry, **which** error is expected to be thrown? You can set the expected error class and message in `assertRejects()`.

Reply:
The error to be thrown in this section is that the size can't exceed 64 Gibs. 

Looking at the documentation of assertRejects, it doesn't have an argument to specify the requirements of the rejection. If another rejection threw, like an invalid name being passed in, then using assertRejects would also say the test passed

Reply:
If another rejection threw, the error message (and possibly class) would be different. Again, `assertRejects()` can check the error message and class thrown. See https://github.com/denoland/std/blob/99200dcd9039eea515f3f850b2f13cf7be585d03/http/server_sent_event_stream_test.ts#L63-L67 for an example.

Also, we should only ever throw `Error`s, and not strings. Please fix any cases where we throw strings instead of `Error`s.

Reply:
> Again, assertRejects() can check the error message and class thrown. See https://github.com/denoland/std/blob/99200dcd9039eea515f3f850b2f13cf7be585d03/http/server_sent_event_stream_test.ts#L63-L67 for an example.

This statement contradicts the documentation of assertRejects https://jsr.io/@std/assert/doc/~/assertRejects